### ISSUE TITLE
Calculation engine result caching

### ISSUE THREAD

**Issue by igitur** [ClosedXML/ClosedXML#667]
cc @Pankraty 

Some issues, e.g. #662 , are about the calculation engine being slow. This could be solved by
- The calculation engine caching a cell's formula's result
- The workbook directly loading and using the cached values that are stored in the file
- The calculation engine implementing the calc chain (although I'm not convinced that the added complexity has any benefit).

My suggestion to fix this is:
- Add optional load parameters to the `XLWorkbook` constructor. Similar to the `SaveOptions` structure that's used by `XLWorkbook.Save()`.
- Add internals to `XLCell` for cached value (or use the existing `ValueCached`?) and a field/property for the date and time the formula was last calculated.
- When cell's result is calculated, store it in the cache field and set date and time property to DateTime.Now.
- When a cell value changes, invalidate all dependent cells - we should decide on whether to invalidate the whole dependency tree onwards.
- Add `IXLWorkbook.RecalculateAllFormulas` and `IXLWorksheet.RecalculateAllFormulas` methods to invalidate and recalculate all cells in applicable scope.
- Add new property in `SaveOptions` to optionally save cached results. Maybe replace the `EvaluateFormulas` property that's already in `SaveOptions`?

More to add?

**Comment by Pankraty** [comment]
Instead of DateTime.Now I would suggest using incremental counter, defined at workbook level. 

- When cells or ranges are modified the counter is increased,
- In XLCell the counter value is stored. When the cell value is obtained, its counter is compared with the workbook state. If they are not equal cell value has to be recalculated. 

With this approach, each cell will be recalculated no more than once after editing, even if the formulas are very complicated.

**Comment by igitur** [comment]
Is that not basically the same approach, except:
- When using a counter, one increments by 1
- When using DateTime, one increments by time passed since last change.

**Comment by Pankraty** [comment]
They are similar indeed, though there is a difference:

- Relying on ```DateTime.Now``` is more fragile since there may be changes performed at the very same moment (multithreading?)
- Comparision of ```int```s (or ```long```s) is performed a few times faster,
- Not mentioning such rare but unpleasant and hard-to-reproduce things like system time changes performed during work (time zone change, time synchronization with NTP server, daylight saving time ending, leap seconds, etc). Time is a complex matter )

Or am I overcomplicating?

**Comment by igitur** [comment]
Valid points, although `ClosedXML` definitely isn't thread-safe yet. Let's assume we continue with a `long` field.

Still, when a cell's counter changes, all its dependent cells' counters will have to change too. Which means we'll have to build a calculation graph to keep track of cell dependencies. And that means we'll have to deal with circular references, like Excel does.

**Comment by Pankraty** [comment]
> Still, when a cell's counter changes, all its dependent cells' counters will have to change too

Not necessarily. Say, we have cells B1and C1 with formula =A1. Once we changed A1 value, its counter became 1, as well as the workbook's counter. B1 and C1 keep holding counter 0. If we try to get the value of B1 it is recalculated and therefore it will have counter=1 too, while C1 remains in the "old" state until we try to get its value. 

I did not dig very deep yet, but it seems to me, ClosedXML already breaks at circular reference occurrence (see ```CellRangeReference```). This part should keep working with the new approach as well.

**Comment by igitur** [comment]
Example:
Upon opening workbook:
A1:  10 (fixed value) (set counter to 0)
B1: =A1 (set counter to 0)
C1: =B1 (set counter to 0)

Change A1 to value 5, and set counter to 1.

Request C1's value. What happens?

**Comment by Pankraty** [comment]
1. Since counter at C1 not equal to workbooks counter we cannot use ValueCached, so perform a recalculation.
2. It leads us to B1, which counter also differs. Keep recalculating.
3. We reach A1 which holds a static value. Use it.
4. B1 recalculation finished, set B1 ValueCached to 5 and counter to 1.
5. C1 recalculation finished, set its ValueCached to 5 and counter to 1 too.

If after that we want to get a value from D1 with the formula ```=MAX(A1:C1)```, we perform a calculation for it. But at this step all cells we need have already been calculated, their ```counter```s are 1, and we simply get a static value from A1 and ValueCached from B1 and C1.

**Comment by igitur** [comment]
Ok, updated example:
Upon opening workbook:
- A1: 10 (fixed value) (set counter to 0)
- B1: =A1 (set counter to 0)
- C1: =B1 (set counter to 0)

- D1: =20 (fixed value) (set counter to 0)
- E1: =D1 (set counter to 0)

Change A1 to value 5, and set counter to 1.

Request E1's value. What happens? Does it re-evaluate through to D1 because the workbook counter has increased to 1 ? This shouldn't happen because the D1/E1 calculation chain was unaffected by the change in value of A1.

**Comment by Pankraty** [comment]
>  Does it re-evaluate through to D1 because the workbook counter has increased to 1 ? 

True. This can be avoided by building a graph of dependencies, so only affected cells were recalculated. This graph should include not only cells but entire ranges too - otherwise such formulas as ```SUM(A:Z)``` would cause an enormous memory allocation. 

Both approaches have their strong sides:

- invalidating dependent cells is preferable in scenarios when reading and writing alternate often (like this would be in a GUI application)
- invalidating entire book and performing recalculation on demand can be faster in scenarios with multiple edits and one recalculation at the end (for example, during saving).

**Comment by igitur** [comment]
Maybe should start with option 2 and think about option 1 (including calculation graph) later. Maintaining a calculation graph is going to very complicated and difficult to "maintain", e.g. when the user deletes some columns.

**Comment by Pankraty** [comment]
@igitur 
I implemented most of the logic in commit 40245e6, but have difficulties with reading cached values from a file, so I think I need your advice on this.
Originally, cached values were both written and read as strings, and there was no need to perform a conversion. If we want to get a strictly-typed value from the file we must know which type the string has to be converted to. But it seems that a type of formula-based cells is neither stored nor determined, it is always Text. 
In other words, do you have ideas how to make tests from a5e6884 pass? Should I modify a saving logic, or a loading logic, or maybe both?

**Comment by igitur** [comment]
You're correct in saying that, internally, ClosedXML stores the values as strings. They are parsed and converted to the correct datatype in the `XLCell.Value` and `XLCell.GetValue()` methods. I haven't looked at your commits yet, but my idea was to implement the caching on the strings directly. So my idea was to abstract out the part where the internally stored string value gets accessed and replace it by the caching algorithms. I'll look at your commits later. Thanks so much for the contributions.

**Comment by Pankraty** [comment]
> Add new property in ```SaveOptions``` to optionally save cached results. Maybe replace the ```EvaluateFormulas``` property that's already in ```SaveOptions```?

After a thought, I see no point in a separate property. If the user chooses ```EvaluateFormulas``` on saving it is presumable he would like to store calculated values in a file. If he does not need calculated values to be saved but only wants a recalculation to be performed, he is free to call ```RecalculateAllFormulas```, without saving. In opposite, if he does not want to recalculate formulas but expects cached values to be saved it also does not make sense, as workbook will contain invalid data in this case.

> my idea was to implement the caching on the strings directly

Yeah, that would be simpler but would affect performance badly. I've already found a single line that took 50% of CPU in my synthetic test:
`if (Double.TryParse(v.ToString(), out tmp))`

Meanwhile, I found a relatively easy way to avoid redundant recalculations without a need to maintain a computation graph. I will explain details later.
The only thing that has to be solved is reading ```ValueCached``` from a file (as I wrote before). What if we just skip it? I mean, ClosedXML can safely write ```ValueCached``` to the file so it would be compatible with the viewers that don't have their own calculation engine. But reading a stored value does not help very much since it is impossible to say whether the saved value is valid or not.
